File: console/console.js
// Copyright 2024 The go-ethereum Authors
// This file is part of the go-ethereum library.
//
// The go-ethereum library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// The go-ethereum library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.

// Pull in a few packages needed for the console
const ethers = require('ethers');
const repl   = require('pretty-repl');
const util   = require('util');

import chalk from 'chalk';

// Connect to the requested node and inject into the repl (ethers)
function dial(url) {
    if (url.startsWith('ws://') || url.startsWith('wss://')) {
        return new ethers.WebSocketProvider(url);
    } else if (url.startsWith('http://') || url.startsWith('https://')) {
        return new ethers.JsonRpcProvider(url);
    } else if (url.startsWith('ipc://')) {
        return new ethers.IpcSocketProvider(url);
    } else {
        throw Error("unknown node type to dial");
    }
}
const api = dial(process.argv[2]);

// Collect all the custom methods we want to expose
const context = {
    ethers: ethers,
    client: api,

    eth: {
        getBlock: async (block, txs = false) => {
            return await api.getBlock(block, txs)
        }
    }
};

// Print some startup headers
const welcome = async () => {
    const client  = await api.send("web3_clientVersion");
    const block   = await api.getBlock();
    const modules = await api.send("rpc_modules");

    console.log(`Welcome to the Tiny Geth console!\n`);

    console.log(`REPL: ${chalk.green("nodejs")} ${chalk.yellow(process.version)}`);
    console.log(`Web3: ${chalk.green("ethers")} ${chalk.yellow("v" + ethers.version)}\n`);

    console.log(`Attached: ${client}`);
    console.log(`At block: ${block.number} (${new Date(1000 * Number(block.timestamp))})`);
    console.log(` Exposed: ${Object.keys(modules).join(' ')}\n`);

    console.log(chalk.grey("• The web3 library uses promises, you need to await appropriately."));
    console.log(chalk.grey("• The usual Geth API methods are exposed to the root namespace."));
    console.log(chalk.grey("• The web3 provider is exposed fully via the `ethers` field."));
    console.log(chalk.grey("• The web3 connection is exposed via the `client` field.\n"));
};

// Start the REPL server and inject all context into it
const startup = async () => {
    const server = repl.start({
        ignoreUndefined: true,
        useGlobal: true,

        prompt: chalk.green('→ '),
        writer: (output) => {
            if (output && typeof output.stack === 'string' && typeof output.message === 'string') {
                return chalk.red(output.stack || output.message);
            }
            return chalk.gray(util.inspect(output, {colors: true, depth: null}));
        }
    });
    Object.assign(server.context, context);
};

welcome().then(() => startup());

File: console/console.go
// Copyright 2024 The go-ethereum Authors
// This file is part of the go-ethereum library.
//
// The go-ethereum library is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// The go-ethereum library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.

package console

import (
	_ "embed"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"

	"github.com/karalabe/tinygeth/console/jsrepl"
	"github.com/karalabe/tinygeth/params"
)

// RunAsProc starts up the console and completely switches over the process to
// it, exiting when the console goes down.
func RunAsProc(url string, eval string, scripts []string) error {
	// Configure the console, cleaning up after
	nodejs, bundle, envs, err := configure(eval, scripts)
	if err != nil {
		return err
	}
	defer os.Remove(bundle)

	// Execute the console, taking over the process space
	return syscall.Exec(nodejs, []string{nodejs, bundle, url}, append(envs, syscall.Environ()...))
}

// RunInProc starts up the console inside the current process, multiplexing the
// standard outputs and consuming the standard input.
func RunInProc(url string, eval string, scripts []string) error {
	// Configure the console, cleaning up after
	nodejs, bundle, envs, err := configure(eval, scripts)
	if err != nil {
		return err
	}
	// Execute the console, keeping it in the current process space
	repl := exec.Command(nodejs, bundle, url)
	repl.Env = append(envs, os.Environ()...)
	repl.Stdin = os.Stdin
	repl.Stdout = os.Stdout
	repl.Stderr = os.Stderr
	return repl.Run()
}

// configure attempts to configure a nodejs based console with a JavaScript REPL
// bundle exported to disk, with some env vars specified.
func configure(eval string, scripts []string) (node string, bundle string, envs []string, err error) {
	// Find the NodeJS executable
	path, err := exec.LookPath("node")
	if err != nil {
		return "", "", nil, err
	}
	path, _ = filepath.Abs(path)

	// Assemble the js repl bundle script
	repl, err := os.CreateTemp("", "tinygeth-console-")
	if err != nil {
		return "", "", nil, err
	}
	defer repl.Close()

	if _, err = repl.WriteString(jsrepl.Bundle + "\n"); err != nil {
		return "", "", nil, err
	}
	if eval == "" {
		// Running in interactive REPL mode
		if _, err = repl.WriteString("module.exports.startup().then(async () => {\n"); err != nil {
			return "", "", nil, err
		}
		for _, script := range scripts {
			blob, err := os.ReadFile(script)
			if err != nil {
				return "", "", nil, err
			}
			if _, err = repl.WriteString(string(blob) + "\n"); err != nil {
				return "", "", nil, err
			}
		}
		if _, err = repl.WriteString("});"); err != nil {
			return "", "", nil, err
		}
	} else {
		// Running in single-shot evaluation mode
		if _, err = repl.WriteString("module.exports.evaluate().then(async () => {\n"); err != nil {
			return "", "", nil, err
		}
		if _, err = repl.WriteString("console.log(await " + eval + "); process.exit();\n"); err != nil {
			return "", "", nil, err
		}
		if _, err = repl.WriteString("});"); err != nil {
			return "", "", nil, err
		}
	}
	// Inject our specific APIs and ENV variables
	envs = []string{
		"TINYGETH_CONSOLE_VERSION=" + params.VersionWithMeta,
	}
	// Return everything for different startups
	return path, repl.Name(), envs, nil
}
